src/usePlaceKit.js

```diff
@@ -4,83 +4,76 @@ import { useEffect, useRef, useState } from 'react';
 import { useStableValue } from './useStableValue';
 
 export const usePlaceKit = (apiKey, options = {}) => {
-  const stableOptions = useStableValue(options);
-
+  // throw error if invalid options
   if (
-    !['object', 'undefined'].includes(typeof stableOptions) ||
-    Array.isArray(stableOptions) ||
-    stableOptions === null
+    typeof options !== 'object' ||
+    Array.isArray(options) ||
+    options === null
   ) {
     throw Error('PlaceKit: `options` parameter is invalid, expected an object.');
   }
 
+  // init states
+  const stableOptions = useStableValue(options);
   const target = useRef(null);
   const [client, setClient] = useState();
   const [state, setState] = useState({
     dirty: false,
     empty: true,
     freeForm: true,
-    hasGeolocation: false,
+    geolocation: false,
   });
 
+  // mount PlaceKit Autocomplete JS
   useEffect(
     () => {
       if (!target.current) {
         return;
       }
 
-      const { handlers, ...opts } = stableOptions || {};
       const pka = placekitAutocomplete(apiKey, {
         target: target.current,
-        ...opts,
-      })
-        .on('open', handlers?.onOpen)
-        .on('close', handlers?.onClose)
-        .on('results', handlers?.onResults)
-        .on('pick', handlers?.onPick)
-        .on('error', handlers?.onError)
-        .on('dirty', handlers?.onDirty)
-        .on('empty', handlers?.onEmpty)
-        .on('freeForm', handlers?.freeForm)
-        .on('state', (newState) => {
-          setState((prev) => ({
-            ...prev,
-            ...newState,
-          }));
-        })
-        .on('geolocation', (bool, pos) => {
-          setState((prev) => ({
-            ...prev,
-            hasGeolocation: bool,
-          }));
-          if (handlers?.onGeolocation) {
-            handlers.onGeolocation(bool, pos);
-          }
-        });
-      setClient(pka);
-
-      // init state
-      setState({
-        ...pka.state,
-        hasGeolocation: pka.hasGeolocation,
       });
+      setClient(pka);
 
       return () => {
         pka.destroy();
         setClient();
       };
     },
-    [apiKey, stableOptions, target.current]
+    [apiKey, target.current]
   );
 
-  const onState = stableOptions.handlers?.onState;
+  // run `pka.configure()` when options update
   useEffect(
     () => {
-      if (onState) {
-        onState(state);
+      if (!client) {
+        return;
       }
+
+      const { handlers, ...opts } = stableOptions || {};
+      client
+        .on('open', handlers?.onOpen)
+        .on('close', handlers?.onClose)
+        .on('results', handlers?.onResults)
+        .on('pick', handlers?.onPick)
+        .on('error', handlers?.onError)
+        .on('dirty', handlers?.onDirty)
+        .on('empty', handlers?.onEmpty)
+        .on('freeForm', handlers?.freeForm)
+        .on('geolocation', handlers?.onGeolocation)
```

Comment:
It's worth running your tests with unstable function references in Strict Mode because I think this code might cause bugs. `useStableValue()` isn't meant to make functions stable. It only works with primitives values, arrays and objects (without function properties).

I would expect these bugs:

- Functions would get turned into factory functions because of the [lazy ref evaluation](https://github.com/placekit/autocomplete-react/blob/7765300244f398a9f8a419392c075ff1e0af3296/src/useStableValue.js#L37)
- It would create a new reference each render (possibly causing infinite loops)

Reply:
Yep, indeed I noticed it, anonymous functions will make the options update each time the parent component re-renders. To address this:
- made Autocomplete JS `configure()` accept all options change, without rerendering the DOM part, see https://github.com/placekit/autocomplete-js/pull/24;
- moved the config in `usePlaceKit` into a separate `useEffect` so that the Autocomplete doesn't remount when option change; computations in `configure()` are quite small, so it's _acceptable_ if options update at each re-render, at least it doesn't reset the whole autocomplete;
- added a section [Avoid re-renders](https://github.com/placekit/autocomplete-react/tree/7765300244f398a9f8a419392c075ff1e0af3296#-avoid-re-renders) in the README to explain that a little bit.

It's way more resilient now, so if someone still use literals in handlers and/or anonymous functions in options, they only get some unnoticeable performance draw from calling `pka.configure()` at each render.